jwst/outlier_detection/outlier_detection_ifu.py

```diff
@@ -53,203 +49,258 @@ def __init__(self, input_models, reffiles=None, **pars):
-    """
-    pass
+            detector_name = uq_det[idet]
+            opt_info = (kern_size[0], kern_size[1], threshold_percent,
+                        diffarr, minarr, normarr, minarr_norm)
+            opt_model = self.create_optional_results_model(opt_info)
+            opt_model.meta.filename = self.make_output_path(
+                basepath=self.input_models.meta.asn_table.products[0].name,
+                suffix=detector_name + '_outlier_output')
+            log.info("Writing out intermediate outlier file {}".format(opt_model.meta.filename))
+            opt_model.save(opt_model.meta.filename)
+
+        del diffarr
+
+        # store some information if the second flagging step is to be done. 
+        if ifu_second_check:
+            # store where the minarr is nan (neighbor pixels have nan so differences produces a nan)
+            nanminarr = np.isnan(minarr)
+            nanindx = np.where(nanminarr)
+
+        # Update DQ flag
+        for i in range(len(self.input_models)):
+            model = self.input_models[i]
+            sci = model.data
+            dq = model.dq
+
+            detector = model.meta.instrument.detector.lower()
+            # only use data from the same detector
+            if detector == uq_det[idet]:
+
+                # There could be a large number of pixels with a sci value of NaN
+                # but the dq flag of DO_NOT_USE has not been set.
+                # This can occur in Non-science regions of the detector.
+                check = np.where(
+                    np.logical_and(~np.bitwise_and(dq, dqflags.pixel['DO_NOT_USE']).astype(bool),
+                                   np.isnan(sci)))
+                log.debug("Number of pixels DQ was not set to DO_NOT_USE and Sci array was Nan{} ".
+                          format(len(check[0])))
+                # set all pixels with dq = DO_NOT_USE to have sci values of Nan
+                bad = np.where(np.bitwise_and(dq, dqflags.pixel['DO_NOT_USE']).astype(bool))
+                sci[bad] = np.nan
+
+                # Basic setting outliers: flagging those at are found in from Percentage cut
+                sci[indx] = np.nan
+                dq[indx] = np.bitwise_or(dq[indx], dqflags.pixel['DO_NOT_USE'])
+                dq[indx] = np.bitwise_or(dq[indx], dqflags.pixel['OUTLIER'])
+
+                nadditional = 0
+                # Second level of setting outliers: flagging pixels were minarr was a Nan
+                # This will also catch pixels that have a sci of Nan but the DQ flags did
+                # not have DO_NOT_USE set
+                if ifu_second_check:
+                    # For counting purposes, count the number of science values that were valid (not Nan)
+                    # after basic flagging in the nanminarr region that will now  be flagged as a Nan.
+                    additional = np.where(~np.isnan(sci[nanindx]))
+                    nadditional = len(additional[0])
```

Comment:
The above two lines can be written as `nadditional = (~np.isnan(sci[nanindx])).sum()`. It's ~4x faster.